### ISSUE TITLE
 java.lang.LinkageError when a tomcat application uses com.google.cloud.bigtable:bigtable-hbase-1.x:jar

### ISSUE THREAD

**Issue by rshivane** [ApptuitAI/JInsight#4]
com.google.cloud.bigtable:bigtable-hbase-1.x:jar uses an older version of io.dropwizard.metrics causing LinkageError in JInsight

Dependency Tree
```
[INFO] +- com.google.cloud.bigtable:bigtable-hbase-1.x:jar:1.0.0-pre3:compile
[INFO] |  \- com.google.cloud.bigtable:bigtable-hbase:jar:1.0.0-pre3:compile
[INFO] |     +- com.google.cloud.bigtable:bigtable-client-core:jar:1.0.0-pre3:compile
[INFO] |     |  +- com.google.cloud.bigtable:bigtable-protos:jar:1.0.0-pre3:compile
[INFO] |     |  |  \- com.google.api.grpc:grpc-google-common-protos:jar:0.1.9:compile
[INFO] |     |  +- commons-logging:commons-logging:jar:1.2:compile
[INFO] |     |  +- com.google.auth:google-auth-library-appengine:jar:0.7.0:compile
[INFO] |     |  +- com.google.api-client:google-api-client:jar:1.22.0:compile
[INFO] |     |  |  \- com.google.oauth-client:google-oauth-client:jar:1.22.0:compile
[INFO] |     |  \- io.dropwizard.metrics:metrics-core:jar:3.1.2:compile
[INFO] |     \- org.apache.hbase:hbase-shaded-client:jar:1.3.1:compile
[INFO] |        +- org.apache.htrace:htrace-core:jar:3.1.0-incubating:compile
[INFO] |        +- com.github.stephenc.findbugs:findbugs-annotations:jar:1.3.9-1:compile
[INFO] |        \- log4j:log4j:jar:1.2.17:compile
```

The error
```
[INFO] [talledLocalContainer] Logback Logger callAppenders exit : java.lang.LinkageError: loader constraint violation: loader (instance of ai/apptuit/metrics/jinsight/ContextualModuleLoader$ModuleClassLoader) previously initiated loading for a different type with name "com/codahale/metrics/MetricRegistry"
[INFO] [talledLocalContainer] Nov 04, 2017 5:31:41 PM org.apache.catalina.core.ApplicationContext log
[INFO] [talledLocalContainer] SEVERE: StandardWrapper.Throwable
[INFO] [talledLocalContainer] org.jboss.byteman.rule.exception.ExecuteException: Logback Logger callAppenders exit  : caught java.lang.LinkageError: loader constraint violation: loader (instance of ai/apptuit/metrics/jinsight/ContextualModuleLoader$ModuleClassLoader) previously initiated loading for a different type with name "com/codahale/metrics/MetricRegistry"
[INFO] [talledLocalContainer]     at org.jboss.byteman.rule.Rule.execute(Rule.java:803)
[INFO] [talledLocalContainer]     at org.jboss.byteman.rule.Rule.execute(Rule.java:740)
[INFO] [talledLocalContainer]     at ch.qos.logback.classic.Logger.callAppenders(Logger.java:266)
[INFO] [talledLocalContainer]     at ch.qos.logback.classic.Logger.buildLoggingEventAndAppend(Logger.java:421)
[INFO] [talledLocalContainer]     at ch.qos.logback.classic.Logger.filterAndLog_0_Or3Plus(Logger.java:383)
[INFO] [talledLocalContainer]     at ch.qos.logback.classic.Logger.info(Logger.java:579)
[INFO] [talledLocalContainer]     at ai.apptuit.ml.web.AnomalyDetectionEngine.init(AnomalyDetectionEngine.java:52)
[INFO] [talledLocalContainer]     at javax.servlet.GenericServlet.init(GenericServlet.java:158)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.StandardWrapper.initServlet(StandardWrapper.java:1183)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.StandardWrapper.loadServlet(StandardWrapper.java:1099)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.StandardWrapper.load(StandardWrapper.java:989)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.StandardContext.loadOnStartup(StandardContext.java:4931)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5241)
[INFO] [talledLocalContainer]     at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:752)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:728)
[INFO] [talledLocalContainer]     at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:734)
[INFO] [talledLocalContainer]     at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:988)
[INFO] [talledLocalContainer]     at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1860)
[INFO] [talledLocalContainer]     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[INFO] [talledLocalContainer]     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[INFO] [talledLocalContainer]     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[INFO] [talledLocalContainer]     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[INFO] [talledLocalContainer]     at java.lang.Thread.run(Thread.java:745)
[INFO] [talledLocalContainer] Caused by: java.lang.LinkageError: loader constraint violation: loader (instance of ai/apptuit/metrics/jinsight/ContextualModuleLoader$ModuleClassLoader) previously initiated loading for a different type with name "com/codahale/metrics/MetricRegistry"
[INFO] [talledLocalContainer]     at ai.apptuit.metrics.jinsight.modules.logback.LogEventTracker.<init>(LogEventTracker.java:53)
[INFO] [talledLocalContainer]     at ai.apptuit.metrics.jinsight.modules.logback.LogEventTracker.<init>(LogEventTracker.java:47)
[INFO] [talledLocalContainer]     at ai.apptuit.metrics.jinsight.modules.logback.LogbackRuleHelper.<clinit>(LogbackRuleHelper.java:30)
[INFO] [talledLocalContainer]     at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[INFO] [talledLocalContainer]     at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[INFO] [talledLocalContainer]     at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[INFO] [talledLocalContainer]     at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
[INFO] [talledLocalContainer]     at org.jboss.byteman.rule.Rule.execute(Rule.java:768)
[INFO] [talledLocalContainer]     ... 23 more
[INFO] [talledLocalContainer] 
```

**Comment by rshivane** [comment]
Ref: http://frankkieviet.blogspot.in/2009/03/javalanglinkageerror-loader-constraint.html

**Comment by rshivane** [comment]
When JInsight modules load dropwizard classes, load them from sys classloader (where the agent loaded the class), instead of loading from app classloader (which by default does not delegate to parent)

Issue fixed in d8cdc79fa48fc150a965c9ffeeeffa12977099cc
